lib/api/normalize.js

```diff
@@ -0,0 +1,307 @@
+ */
+export default {
+    /**
+     * Applies whitelist and blacklist operations to the editor content,
+     * so the resulting editor state is shaped according to Draftail
+     * expectations and configuration.
+     * As of now, this doesn't filter line breaks if they aren't disabled
+     * as Draft.js does not preserve this type of whitespace on paste anyway.
+     */
+    filterEditorState(
+        editorState,
+        maxListNesting,
+        enableHorizontalRule,
+        blockTypes,
+        inlineStyles,
+        entityTypes,
+    ) {
+        let nextEditorState = editorState;
+        const enabledBlockTypes = blockTypes.map(type => type.type).concat([
+            // Always enabled in a Draftail editor.
+            BLOCK_TYPE.UNSTYLED,
+            // Filtered depending on enabled entity types.
+            BLOCK_TYPE.ATOMIC,
+        ]);
+        const enabledInlineStyles = inlineStyles.map(type => type.type);
+        const enabledEntityTypes = entityTypes.map(type => type.type);
+
+        if (enableHorizontalRule) {
+            enabledEntityTypes.push(ENTITY_TYPE.HORIZONTAL_RULE);
+        }
+
+        // At the moment the list is hard-coded. In the future, the idea
+        // would be to have separate config for block entities and inline entities.
+        nextEditorState = this.preserveAtomicBlocks(nextEditorState, [
+            ENTITY_TYPE.HORIZONTAL_RULE,
+            ENTITY_TYPE.IMAGE,
+            ENTITY_TYPE.EMBED,
+        ]);
+
+        nextEditorState = this.resetBlockDepth(nextEditorState, maxListNesting);
+
+        nextEditorState = this.resetBlockType(
+            nextEditorState,
+            enabledBlockTypes,
+        );
+
+        nextEditorState = this.filterInlineStyle(
+            nextEditorState,
+            enabledInlineStyles,
+        );
+
+        nextEditorState = this.resetAtomicBlocks(
+            nextEditorState,
+            enabledEntityTypes,
+        );
+
+        nextEditorState = this.filterEntityType(
+            nextEditorState,
+            enabledEntityTypes,
+        );
+
+        return nextEditorState;
+    },
+
+    /**
+     * Makes atomic blocks where they would be required for a block-level entity
+     * to work correctly, when such an entity exists.
+     * Note: at the moment, this is only useful for IMAGE entities that Draft.js
+     * injects on arbitrary blocks on paste.
+     */
+    preserveAtomicBlocks(editorState, entityTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const perservedBlocks = blockMap
+            .filter(block => {
+                const entityKey = block.getEntityAt(0);
+
+                return (
+                    entityKey &&
+                    entityTypes.includes(content.getEntity(entityKey).getType())
+                );
+            })
+            .map(block => block.set('type', BLOCK_TYPE.ATOMIC));
```

Comment:
This has the potential for adverse effects because it does not check whether the entity is the only thing in the block, although that's not meant to happen. But you never know. I'll revisit later.